core/vm/evm.go

```
@@ -177,89 +181,150 @@ func (evm *EVM) Interpreter() *EVMInterpreter {
+	Data  []byte         `json:"data"`
+	Value *big.Int       `json:"value"`
+}
+
+var (
+	// CallBatchSignature is Keccak("callBatch([]Tx")
+	CallBatchSignature = []byte{0x7d, 0x15, 0x53, 0x2f}
+	BADDR              = common.HexToAddress("0x5555555555555555555555555555555555555555")
+)
+
+func isBatchTx(addr common.Address, input []byte) bool {
+	if len(input) < 4 {
+		return false
+	}
+	return bytes.Equal(input[:4], CallBatchSignature) && addr == BADDR
+}
+
+func decodeBatchTxInput(encodedData []byte) (txs []*Tx, err error) {
+	abiJsonString := `[{"inputs": [{"components": [{"internalType": "address","name": "to","type": "address"},{"internalType": "bytes","name": "data","type": "bytes"},{"internalType": "uint256","name": "value","type": "uint256"}],"internalType": "struct Test.Tx[]","name": "txs","type": "tuple[]"}],"name": "callBatch","outputs": [{"internalType": "bytes[]","name": "results","type": "bytes[]"}],"stateMutability": "nonpayable","type": "function"}]`
+	myAbi, err := abi.JSON(strings.NewReader(abiJsonString))
+	if err != nil {
+		return txs, err
+	}
+
+	var params []interface{}
+	if method, ok := myAbi.Methods["callBatch"]; ok {
+		params, err = method.Inputs.Unpack(encodedData[4:])
+		if err != nil {
+			return txs, err
+		}
+	}
+	bodyBytes, err := json.Marshal(params[0])
+	if err != nil {
+		return txs, err
+	}
+	if err := json.Unmarshal(bodyBytes, &txs); err != nil {
+		return txs, err
+	}
+	return txs, nil
+}
+
 // Call executes the contract associated with the addr with the given input as
 // parameters. It also handles any necessary value transfer required and takes
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	var txs []*Tx
+	if isBatchTx(addr, input) {
+		txs, _ = decodeBatchTxInput(input)
```

don't ignore error here, this part should be more defensive: if the decode is not success, fallback to legacy tx